File: src/eQuantic.UI.Runtime/src/shared/components/CodeLanguages.ts
import { CSharpLanguage, JsonLanguage, PlainTextLanguage, PythonLanguage, TypeScriptLanguage, XmlLanguage } from "../runtime-exports";

export class CodeLanguages {
    static _cSharp: any | undefined;

    static get cSharp(): any {
        return CodeLanguages._cSharp ??= new CSharpLanguage();
    }

    static _typeScript: any | undefined;

    static get typeScript(): any {
        return CodeLanguages._typeScript ??= new TypeScriptLanguage();
    }

    static _python: any | undefined;

    static get python(): any {
        return CodeLanguages._python ??= new PythonLanguage();
    }

    static _json: any | undefined;

    static get json(): any {
        return CodeLanguages._json ??= new JsonLanguage();
    }

    static _xml: any | undefined;

    static get xml(): any {
        return CodeLanguages._xml ??= new XmlLanguage();
    }

    static _plainText: any | undefined;

    static get plainText(): any {
        return CodeLanguages._plainText ??= new PlainTextLanguage();
    }

    static _known: Record<string, any> | undefined;

    static get known(): Record<string, any> {
        return CodeLanguages._known ??= { ['c#']: CodeLanguages.cSharp, ['csharp']: CodeLanguages.cSharp, ['cs']: CodeLanguages.cSharp, ['typescript']: CodeLanguages.typeScript, ['ts']: CodeLanguages.typeScript, ['javascript']: CodeLanguages.typeScript, ['js']: CodeLanguages.typeScript, ['tsx']: CodeLanguages.typeScript, ['jsx']: CodeLanguages.typeScript, ['python']: CodeLanguages.python, ['py']: CodeLanguages.python, ['json']: CodeLanguages.json, ['xml']: CodeLanguages.xml, ['csproj']: CodeLanguages.xml, ['html']: CodeLanguages.xml, ['plist']: CodeLanguages.xml, ['text']: CodeLanguages.plainText, ['txt']: CodeLanguages.plainText, ['plain']: CodeLanguages.plainText };
    }

    static register(name: string, language: any) {
        return CodeLanguages.known[name] = language;
    }

    static for(name: string | null) {
        let language: any; if ((!name || !name.trim())) return CodeLanguages.plainText;
        let key = (_s => { const _c = '.'; let _i = 0; while (_i < _s.length && _c.includes(_s[_i])) _i++; return _s.slice(_i); })(name);
        return (Object.prototype.hasOwnProperty.call(CodeLanguages.known, key) ? ((language = CodeLanguages.known[key]), true) : false) ? language : CodeLanguages.plainText;
    }
}

File: src/eQuantic.UI.Code/Languages/CodeLanguages.cs
namespace eQuantic.UI.Code;

/// <summary>
/// The languages this SDK ships, by name. A registry rather than an enum because an app may bring
/// its own: <c>CodeLanguages.Register(new SqlLanguage())</c> and the editor colours it with no
/// change here.
/// </summary>
public static class CodeLanguages
{
    public static readonly ICodeLanguage CSharp = new CSharpLanguage();
    public static readonly ICodeLanguage TypeScript = new TypeScriptLanguage();
    public static readonly ICodeLanguage Python = new PythonLanguage();
    public static readonly ICodeLanguage Json = new JsonLanguage();
    public static readonly ICodeLanguage Xml = new XmlLanguage();
    public static readonly ICodeLanguage PlainText = new PlainTextLanguage();

    private static readonly Dictionary<string, ICodeLanguage> Known = new(StringComparer.OrdinalIgnoreCase)
    {
        ["c#"] = CSharp, ["csharp"] = CSharp, ["cs"] = CSharp,
        ["typescript"] = TypeScript, ["ts"] = TypeScript,
        ["javascript"] = TypeScript, ["js"] = TypeScript, ["tsx"] = TypeScript, ["jsx"] = TypeScript,
        ["python"] = Python, ["py"] = Python,
        ["json"] = Json,
        ["xml"] = Xml, ["csproj"] = Xml, ["html"] = Xml, ["plist"] = Xml,
        ["text"] = PlainText, ["txt"] = PlainText, ["plain"] = PlainText,
    };

    /// <summary>Adds (or replaces) a language under a name — an app's own dialect joins here.</summary>
    public static void Register(string name, ICodeLanguage language) => Known[name] = language;

    /// <summary>The language for a name or a file extension; plain text when nothing matches, which
    /// is the answer that renders rather than the one that throws.</summary>
    public static ICodeLanguage For(string? name)
    {
        if (string.IsNullOrWhiteSpace(name)) return PlainText;
        var key = name.TrimStart('.');
        return Known.TryGetValue(key, out var language) ? language : PlainText;
    }
}
